addons/sms_twilio/tests/test_sms_twilio.py

```diff
@@ -0,0 +1,253 @@
+                "price": None,
+                "price_unit": "USD",
+                "sid": "SMfake",
+                "status": "queued",
+                "subresource_uris": {
+                    "media": "/2010-04-01/Accounts/ACfake/Messages/SMfake/Media.json"
+                },
+                "to": to_number,
+                "uri": "/2010-04-01/Accounts/ACfake/Messages/SMfake.json",
+            }
+            nok_json = {
+                'code': 21211,
+                'message': "Invalid 'To' Phone Number: +324863XXXX",
+                'more_info': 'https://www.twilio.com/docs/errors/21211',
+                'status': 400,
+            }
+            if to_number == self.valid_phone_number:
+                response = MockResponse(json_data=ok_json, status_code=201)
+            else:
+                response = MockResponse(json_data=nok_json, status_code=400)
+            return response
+
+        self._clear_sms_sent()
+        sms_create_origin = SmsSms.create
+
+        def _sms_sms_create(model, *args, **kwargs):
+            res = sms_create_origin(model, *args, **kwargs)
+            self._new_sms += res.sudo()
+            return res
+
+        def _validate_twilio_signature_mocked(model, *args, **kwargs):
+            return True
+
+        try:
+            with (
+                patch.object(SmsSms, 'create', autospec=True, wraps=SmsSms, side_effect=_sms_sms_create),
+                patch.object(SmsTwilioApi, '_sms_twilio_send_request', autospec=True, side_effect=_sms_twilio_send_request),
+                patch.object(SmsTwilioController, '_validate_twilio_signature', autospec=True, side_effect=_validate_twilio_signature_mocked),
+            ):
+                self._sms += [{
+                    'number': phone_number,
+                    'body': sms_content,
+                }]
+                self._new_sms += self.env['sms.sms'].sudo().create(self._sms[-1])
+                yield
+        finally:
+            pass
+
+
+@tagged('post_install', '-at_install')
+class TestSmsTwilio(MockSmsTwilioApi, SMSCase, common.TransactionCase):
+    def test_send_to_valid_number(self):
+        with self.mock_sms_twilio_gateway(self.valid_phone_number, "Valid phone number"):
+            self._new_sms[-1].send(unlink_failed=True, unlink_sent=True, raise_exception=True)
+            self.assertTrue(self._new_sms[-1].to_delete)
+            self.assertFalse(self._new_sms[-1].failure_type)
+            self.assertEqual(self._new_sms[-1].body, "Valid phone number")
+
+    def test_send_to_invalid_number(self):
+        with self.mock_sms_twilio_gateway(self.invalid_phone_number, "Invalid phone number"):
+            self._new_sms[-1].send(unlink_failed=True, unlink_sent=True, raise_exception=False)
+            self.assertTrue(self._new_sms[-1].to_delete)
+            self.assertEqual(self._new_sms[-1].failure_type, "sms_number_format")
+            self.assertEqual(self._new_sms[-1].body, "Invalid phone number")
```

Comment:
fyi there's utils like `assertSMS` for this that's supposed to give nicer results.

Reply:
Indeed